### PR TITLE
Add Settings Types to PlayerSettings Debug Tool Window

### PR DESCRIPTION
**Pull request by HiRezDanxor**

### CODE REVIEW

File: RallyHereDebugTool/Source/Private/RHDTW_PlayerSettings.cpp [hunk 1]
```diff
@@ -144,6 +150,69 @@ void FRHDTW_PlayerSettings::DoModifySettings()
...
+
+	if (!pRH_LocalPlayerSubsystem->GetAuthContext()->IsLoggedIn())
+	{
+		ImGui::Text("Requesting Types requires a logged in local player for Auth Context");
+		return;
+	}
+
+	if (ImGui::Button("Get Setting Types"))
+	{
+		typedef RallyHereAPI::Traits_GetConfigForAllSettingTypes TGetConfigSettings;
```
Comment (HiRezDanxor) [inline RallyHereDebugTool/Source/Private/RHDTW_PlayerSettings.cpp:177]: I just put the call in here and cached the data on the debug tool itself, I thought about putting it somewhere in the Integration Layer but was not sure where was a good spot, as Settings is kinda a subpiece of PlayerInfos which there are many of, so felt a bit weird on them, up for suggestions

Comment (HiRezDanxor) [reply RallyHereDebugTool/Source/Private/RHDTW_PlayerSettings.cpp:177]: Are there Catalog Templates? What uses them?

Session Type Templates are cached in various ways on the URH_MatchmakingBrowserCache I believe.  

The problem is Settings in general isn't its own Class, so it doesn't really have a nice home for code.  Which may be the solution or we just leave this call where I put it

Comment (nonathaj) [reply RallyHereDebugTool/Source/Private/RHDTW_PlayerSettings.cpp:177]: There are no catalog templates, but all the components of catalog (loot, item, vendor, etc) are pieces of configuration from the dev portal that clients like to have a view of.  Session templates and Setting Types are a similar kind of data.  Stuff that the game teams configure in the Dev Portal/API, and may want their clients to pull a copy of that down to do client side validation against.

I'm in favor of giving it its own class.  Even if we only use it to cache this configuration data.

File: RallyHereDebugTool/Source/Private/RHDTW_PlayerSettings.cpp [hunk 2]
```diff
@@ -144,6 +152,61 @@ void FRHDTW_PlayerSettings::DoModifySettings()
...
+
+	if (pRH_SettingsSubsystem->GetCachedSettingTypes().Num() > 0)
+	{
+		for (const auto& Pair : pRH_SettingsSubsystem->GetCachedSettingTypes())
+		{
+			if (ImGui::TreeNodeEx(TCHAR_TO_UTF8(*Pair.Key), RH_DefaultTreeFlagsDefaultOpen))
+			{
+				for (const auto& VersionPair : Pair.Value.GetVersions())
+				{
+					if (ImGui::TreeNodeEx(TCHAR_TO_UTF8(*FString::Printf(TEXT("%s##%s"), *VersionPair.Key, *Pair.Key)), RH_DefaultTreeFlagsLeaf | ImGuiTreeNodeFlags_DefaultOpen))
```
Comment (hirez-cranney) [inline RallyHereDebugTool/Source/Private/RHDTW_PlayerSettings.cpp:192]: Anh added a ImGuiDisplayJsonObject() utility function for displaying JSON blobs consistently.

File: RallyHereIntegration/Source/RallyHereIntegration/Private/RH_SettingsSubsystem.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+void URH_SettingsSubsystem::GetSettingTypes(const FRH_GenericSuccessBlock& Delegate)
+{
+	UE_LOG(LogRallyHereIntegration, VeryVerbose, TEXT("[%s]"), ANSI_TO_TCHAR(__FUNCTION__));
+
+	auto Request = TGetConfigSettings::Request();
+	Request.AuthContext = GetAuthContext();
+
+	if (!TGetConfigSettings::DoCall(RH_APIs::GetSettingsAPI(), Request, TGetConfigSettings::Delegate::CreateLambda([this, Delegate](const TGetConfigSettings::Response& Resp)
+		{
+			CachedSettingsTypes.Empty();
```
Comment (hirez-cranney) [inline RallyHereIntegration/Source/RallyHereIntegration/Private/RH_SettingsSubsystem.cpp:17]: This needs to do some validation checking (IsSuccessful, etc).

Would suggest looking into the FRH_SimpleQueryHelper class (config subsystem is a good example).  It takes two lambas, one for the update logic (only triggered on success), and one for completion (with error information, and uses FRH_GenericSuccessBlockWithError).  Would suggest converting this function itself to use the FRH_GenericSuccessBlockWithError, as I have been slowly converting things to use it as it provides more information about why something failed.

File: RallyHereIntegration/Source/RallyHereIntegration/Private/RH_SettingsSubsystem.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+{
+	UE_LOG(LogRallyHereIntegration, VeryVerbose, TEXT("[%s]"), ANSI_TO_TCHAR(__FUNCTION__));
+
+	auto Request = TGetConfigSettings::Request();
+	Request.AuthContext = GetAuthContext();
+
+	auto Helper = MakeShared<FRH_SimpleQueryHelper<TGetConfigSettings>>(
+		TGetConfigSettings::Delegate::CreateLambda([this, Delegate](const TGetConfigSettings::Response& Resp)
+			{
+				if (Resp.IsSuccessful())
```
Comment (hirez-cranney) [inline RallyHereIntegration/Source/RallyHereIntegration/Private/RH_SettingsSubsystem.cpp:18]: Existing check does not hurt, but the first delegate here is only called if Resp.IsSuccessful() is true. Does not get called for semi-succesful non-updates (ex: NOT_MODIFIED codes).

The second delegate (completion delegate) always gets called.

**Review by hirez-cranney — APPROVED** [review]

**Pull request merged**